File: LegacyBite/Controllers/Scanner/ScannerViewController.m
//
//  ScannerViewController.m
//  LegacyBite
//
//  Created by Grizly on 15.07.26.
//

#import "ScannerViewController.h"
#import "AVFoundation/AVFoundation.h"
#import "CameraViewController.h"
#import "ProductManager.h"
#import "ProductCardViewController.h"
#import "LegacyBite-Swift.h"

@interface ScannerViewController () <CameraViewControllerDelegate>

@property (strong, nonatomic) ScannerViewModel * viewModel;

@end

@implementation ScannerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self initViewModel];
}

-(void)initViewModel{
    self.viewModel = [[ScannerViewModel alloc] init];
    __weak typeof(self) weakSelf = self;
    
    self.viewModel.onLoadingChanged = ^(BOOL isLoading){
        dispatch_async(dispatch_get_main_queue(), ^{
            if (isLoading) {
                [weakSelf showLoaderWithBlocksInteraction:NO];
            } else {
                [weakSelf hideLoader];
            }
        });
    };
    
    self.viewModel.onProductLoaded = ^(SSProductObject * product){
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf showProduct:product];
        });
        
    };
    
    self.viewModel.onError = ^(NSError * error){
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf showAlertWith:error.localizedDescription];
        });
    };
}

-(IBAction)scanBarcodeAction:(id)sender{
    AVAuthorizationStatus status = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    if(status == AVAuthorizationStatusAuthorized){
        [self presentCameraViewController];
    } else if (status == AVAuthorizationStatusNotDetermined){
        [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (granted) {
                    [self presentCameraViewController];
                } else {
                    [self showAccessDeniedAlert];
                }
            });
        }];
    } else {
        [self showAccessDeniedAlert];
    }
}

-(void)presentCameraViewController{
    UINavigationController * nc = [[UIStoryboard storyboardWithName:@"Main" bundle:nil]instantiateViewControllerWithIdentifier:@"CameraNavController"];
    CameraViewController * vc = (CameraViewController *)nc.viewControllers.firstObject;
    vc.delegate = self;
    nc.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:nc animated:false completion:nil];
}

-(void)showAccessDeniedAlert{
    
    NSString * title = @"";
    NSString * message = @"Enable camera access in Settings to scan product barcodes.";
    
    UIAlertController * alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction * toSettingsAction = [UIAlertAction actionWithTitle:@"Settings" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        NSURL * url = [NSURL URLWithString:UIApplicationOpenSettingsURLString];
        if([[UIApplication sharedApplication] canOpenURL:url]){
            [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
        }
    }];
    UIAlertAction * cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil];

        
    [alert addAction:toSettingsAction];
    [alert addAction:cancelAction];
    [self presentViewController:alert animated:true completion:nil];
}

-(void)showAlertWith:(NSString *)alertMessage{
    UIAlertController * alert = [UIAlertController alertControllerWithTitle:@"Error" message:alertMessage preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction * ok = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:ok];
    [self presentViewController:alert animated:true completion:nil];
}

-(void)showProduct:(SSProductObject *)product{
    ProductCardViewController * vc = [[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:@"ProductCardViewController"];
    vc.product = product;
    [self.navigationController pushViewController:vc animated:true];
}

#pragma mark: CameraViewControllerDelegate
-(void)didScanBarCode:(NSString *)barCode{
    [self.viewModel loadProductWithBarCode:barCode];
}

@end

File: LegacyBite/Controllers/Scanner/ScannerViewController.swift
//
//  ScannerViewController.swift
//  LegacyBite
//
//  Created by Grizly on 24.07.26.
//

import UIKit
import AVFoundation
import Combine

class ScannerViewController: UIViewController {

    var hasCameraAccess: Bool{
        get async {
            let status = AVCaptureDevice.authorizationStatus(for: .video)
            var isAuthorized = status == .authorized
            if status == .notDetermined {
                isAuthorized = await AVCaptureDevice.requestAccess(for: .video)
            }
            
            return isAuthorized
        }
    }
    
    let viewModel = ScannerViewModel()
    
    private var cancellables: Set<AnyCancellable> = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        bindViewModel()
    }
    
    private func bindViewModel(){
        viewModel.$state.receive(on: DispatchQueue.main).sink{ [weak self] state in
            self?.render(state)
        }.store(in: &cancellables)
    }
    private func render(_ state: ScannerViewModel.DataState) {
        switch state {
        case .idle:
            hideLoader()
            
        case .loading:
            showLoader()
            
        case .failed(let message):
            hideLoader()
            showAlert(message: message)
  
        case .loaded(let product):
            hideLoader()
            showProduct(product)
        }
    }
    
    @IBAction func scanBarCode(_ sender: Any) {
        Task {@MainActor in
            let authorized = await hasCameraAccess
            if authorized {
                presentCameraViewController()
            } else {
                showAccessDeniedAlert()
            }
        }
    }
    
    private func presentCameraViewController() {
        let storyboard = UIStoryboard(name: "Main", bundle: nil)

        guard
            let navigationController = storyboard.instantiateViewController( withIdentifier: "CameraNavController" ) as? UINavigationController, let cameraViewController = navigationController.viewControllers.first as? CameraViewController
        else {
            assertionFailure("CameraNavController is configured incorrectly")
            return
        }

        cameraViewController.delegate = self
        navigationController.modalPresentationStyle = .fullScreen

        present(navigationController, animated: false)
    }
    
    private func showAccessDeniedAlert() {
        let alert = UIAlertController(
            title: nil, message: "Enable camera access in Settings to scan product barcodes.", preferredStyle: .alert
        )

        let settingsAction = UIAlertAction( title: "Settings", style: .default
        ) { _ in
            guard let url = URL(string: UIApplication.openSettingsURLString) else {
                return
            }
            UIApplication.shared.open(url)
        }

        let cancelAction = UIAlertAction(
            title: "Cancel",
            style: .cancel
        )

        alert.addAction(settingsAction)
        alert.addAction(cancelAction)

        present(alert, animated: true)
    }

    private func showAlert(message: String) {
        let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)

        alert.addAction(
            UIAlertAction( title: "OK", style: .default )
        )

        present(alert, animated: true)
    }
    
    private func showProduct(_ product: SSProductObject) {
        let storyboard = UIStoryboard(name: "Main", bundle: nil)

        guard let viewController = storyboard.instantiateViewController(
            withIdentifier: "ProductCardViewController"
        ) as? ProductCardViewController else {
            assertionFailure("ProductCardViewController not found")
            return
        }

        viewController.product = product
        navigationController?.pushViewController(viewController, animated: true)
    }
    
}

extension ScannerViewController : @MainActor CameraViewControllerDelegate{
    func didScanBarCode(_ barCode: String) {
        Task {
            await viewModel.loadProduct(barCode: barCode)
        }
    }
    
    
}
